File: qa/verify-text-fitting.py
"""Independent native XLS reopen (xlrd) and optional rendered PDF checks.
Usage: bundled-python qa/verify-text-fitting.py /path/to/fixture-directory
Generate fixtures with qa/verify-text-fitting.ts first. Render general-full-fit.xls
and student-full-fit.xls to PDFs with the bundled override soffice and original
fonts, each with a unique temporary profile. No Office GUI is used.
"""
import json
from pathlib import Path
import sys
import xlrd

root=Path(sys.argv[1])
fixtures=json.loads((root/'expected.json').read_text())
cell_count=0
for fixture in fixtures:
    book=xlrd.open_workbook(root/fixture['file'],formatting_info=True)
    sheet=book.sheet_by_name(fixture['sheet'])
    for expected in fixture['cells']:
        value=expected['value']
        text=value.get('text') if isinstance(value,dict) else '' if value is None else value
        cell=sheet.cell(expected['row'],expected['col'])
        assert cell.value==text,(fixture['file'],expected['address'],cell.value,text)
        style=book.xf_list[cell.xf_index]
        assert cell.xf_index==expected['style']
        assert bool(style.alignment.text_wrapped)==expected['wrap']
        assert book.font_list[style.font_index].height/20==expected['size']
        if isinstance(value,dict) and value.get('runs'):
            assert sheet.rich_text_runlist_map[expected['row'],expected['col']]==[(run['start'],run['fontIndex']) for run in value['runs']]
        cell_count+=1

pdf_results=[]
for kind,pages in [('general',2),('student',1)]:
    file=root/f'{kind}-full-fit.pdf'
    if not file.exists():
        continue
    import pdfplumber
    fixture=next(item for item in fixtures if item['file']==f'{kind}-full-fit.xls')
    with pdfplumber.open(file) as pdf:
        assert len(pdf.pages)==pages
        page=pdf.pages[0]
        assert abs(page.width-595.28)<1 and abs(page.height-841.89)<1
        raw=''.join(char['text'] for char in page.chars)
        raw_chars=[char for char in page.chars for _ in char['text']]
        for expected in fixture['cells']:
            if expected['address'] not in ['C9','C14','C26']:
                continue
            value=expected['value']
            text=value.get('text') if isinstance(value,dict) else value or ''
            start=raw.index(str(text)[:5])
            first=raw_chars[start]
            bottom=min((edge for edge in page.edges if edge.get('orientation')=='h' and edge['x0']<=first['x0'] and edge['x1']>first['x0']+20 and edge['top']>first['top']+2),key=lambda edge:edge['top'])
            cropped=page.crop((first['x0']-1,first['top']-2,bottom['x1'],bottom['top']+0.1))
            extracted=''.join(cropped.extract_text().split())
            assert ''.join(str(text).split()) in extracted,(kind,expected['address'],'rendered text missing')
        assert all(0<=char['x0']<=char['x1']<=page.width and 0<=char['top']<=char['bottom']<=page.height for char in page.chars)
        pdf_results.append({'kind':kind,'pages':pages,'mainPage':'A4 portrait single page','purposeWorkNotes':'complete','allGlyphsInsidePage':True})
result={'nativeFiles':len(fixtures),'exactCellsWithFontWrapAndRichText':cell_count,'pdfChecks':pdf_results}
(root/'verification.json').write_text(json.dumps(result,ensure_ascii=False,indent=2)+'\n')
print(json.dumps(result,ensure_ascii=False,indent=2))

File: qa/verify-text-fitting.ts
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { fitClaimText } from '../lib/xls/text-fitting';
import { patchXls } from '../lib/xls/xls-exporter';
import { buildClaimChanges } from '../lib/xls/claim-mapping';
import { createSampleDraft, calculateClaim } from '../lib/claim/claim-engine';
import type { TemplateManifest } from '../lib/xls/template-manifest';
const out = process.argv[2] || 'qa-output/text-fitting';
mkdirSync(out, { recursive: true });
const expected = [];
for (const kind of ['general', 'student'] as const)
  for (let count = 1; count <= 12; count++) {
    const m = JSON.parse(
      readFileSync(`public/templates/${kind}-${count}.json`, 'utf8'),
    ) as TemplateManifest;
    const changes = {
      [m.sheetName]: {
        [m.segments[0].fields.location]: '美國\n波士頓\n研究所',
      },
    };
    const fit = fitClaimText(m, changes);
    if (fit.issues.length) throw new Error(JSON.stringify(fit.issues));
    const file = `${kind}-${count}-fit.xls`;
    writeFileSync(
      `${out}/${file}`,
      patchXls(
        readFileSync(`public/templates/${m.file}`),
        fit.changes,
        fit.formatting,
      ),
    );
    expected.push({
      file,
      sheet: m.sheetName,
      cells: Object.entries(fit.changes[m.sheetName]).map(
        ([address, value]) => {
          const cell = fit.manifest.sheets[0].cells.find(
            (c) => c.address === address,
          )!;
          const style = fit.manifest.styles[cell.styleIndex];
          return {
            address,
            row: cell.row,
            col: cell.col,
            value,
            style: cell.styleIndex,
            wrap: style.wrap,
            size: fit.manifest.fonts[style.fontIndex].size,
          };
        },
      ),
    });
  }
for (const kind of ['general', 'student'] as const) {
  const m = JSON.parse(
    readFileSync(`public/templates/${kind}-4.json`, 'utf8'),
  ) as TemplateManifest;
  const d = createSampleDraft();
  d.template = kind;
  d.purpose =
    '國際學術研討會發表研究成果並參與跨國研究合作與方法交流座談會議'.repeat(2);
  d.notes = '出差經費分攤依補助核定項目辦理。'.repeat(
    kind === 'general' ? 5 : 3,
  );
  d.days = d.days.map((day) => ({
    ...day,
    work: '第一行工作記要\n第二行研究討論',
  }));
  const fit = fitClaimText(m, buildClaimChanges(d, calculateClaim(d), m));
  if (fit.issues.length) throw new Error(JSON.stringify(fit.issues));
  const file = `${kind}-full-fit.xls`;
  writeFileSync(
    `${out}/${file}`,
    patchXls(
      readFileSync(`public/templates/${m.file}`),
      fit.changes,
      fit.formatting,
    ),
  );
  expected.push({
    file,
    sheet: m.sheetName,
    cells: Object.entries(fit.changes[m.sheetName]).map(([address, value]) => {
      const cell = fit.manifest.sheets[0].cells.find(
        (c) => c.address === address,
      )!;
      const style = fit.manifest.styles[cell.styleIndex];
      return {
        address,
        row: cell.row,
        col: cell.col,
        value,
        style: cell.styleIndex,
        wrap: style.wrap,
        size: fit.manifest.fonts[style.fontIndex].size,
      };
    }),
  });
}
writeFileSync(`${out}/expected.json`, JSON.stringify(expected, null, 2));
console.log(`Wrote ${expected.length} native XLS fitting fixtures.`);
